server/src/main/java/org/elasticsearch/cluster/coordination/ElectionStrategy.java

```diff
@@ -105,10 +110,14 @@ public void beforeCommit(long term, long version, ActionListener<Void> listener)
         listener.onResponse(null);
     }
 
-    public boolean nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
+    public NodeEligibility nodeMayWinElection(ClusterState lastAcceptedState, DiscoveryNode node) {
         final String nodeId = node.getId();
-        return lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
+        if (lastAcceptedState.getLastCommittedConfiguration().getNodeIds().contains(nodeId)
             || lastAcceptedState.getLastAcceptedConfiguration().getNodeIds().contains(nodeId)
-            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId));
+            || lastAcceptedState.getVotingConfigExclusions().stream().noneMatch(vce -> vce.getNodeId().equals(nodeId))) {
+            return new NodeEligibility(true, "");
+        }
+
+        return new NodeEligibility(false, "node is ineligible for election, not part of the voting configuration");
```

Comment:
I couldn't figure out what the difference between LastAccepted and LastCommitted is, or really what VotingConfiguration represents -- are all master nodes listed in VotingConfiguration??

@DaveCTurner

Reply:
Voting configuration is described here: https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery-voting.html

Reply:
Also it might not be necessary, to come up with a reason for each of these. I think, it would be enough to just add the shutdown-related message, since the rest is not necessary for Stateful. We could expand it in a follow up.